packages/library-base/src/commonMain/kotlin/io/realm/RealmQuery.kt

```diff
@@ -0,0 +1,361 @@
+//     *
+//     * This function returns the average with the same precision provided by the native database.
+//     *
+//     * @param property the property on which the result will be computed.
+//     * @return a [RealmScalarQuery] returning the average value for the given property represented
+//     * as a [Double].
+//     * @throws [IllegalArgumentException] if the property is not a [Number].
+//     */
+//    fun average(property: String): RealmScalarQuery<Double>
+
+    /**
+     * Returns a [RealmScalarQuery] that counts the number of objects that fulfill the query
+     * conditions.
+     *
+     * @return a [RealmScalarQuery] that counts the number of objects for a given query.
+     */
+    fun count(): RealmScalarQuery<Long>
+}
+
+/**
+ * Query returning [RealmResults].
+ */
+interface RealmElementQuery<E> : Flowable<RealmResults<E>> {
+
+    /**
+     * Finds all objects that fulfill the query conditions and returns them in a blocking fashion.
+     *
+     * Launching heavy queries from the UI thread may result in a drop of frames or even ANRs. **We
+     * do not recommend doing so.** If you want to prevent these behaviors you can obtain the
+     * results asynchronously using [asFlow] instead.
+     *
+     * @return a [RealmResults] instance containing matching objects. If no objects match the
+     * condition, an instance with zero objects is returned.
+     */
+    fun find(): RealmResults<E>
+
+    /**
+     * Finds all objects that fulfill the query conditions and returns them asynchronously as a
+     * [Flow].
+     *
+     * If there is any changes to the objects represented by the query backing the [RealmResults],
+     * the flow will emit the updated results. The flow will continue running indefinitely until
+     * canceled.
+     *
+     * The change calculations will run on the thread represented by
+     * [RealmConfiguration.Builder.notificationDispatcher].
+     *
+     * @return a flow representing changes to the [RealmResults] resulting from running this query.
+     */
+    override fun asFlow(): Flow<RealmResults<E>>
+}
+
+/**
+ * Query returning a single [RealmObject].
+ *
+ * TODO : The interaction with primitive queries might be a bit awkward
+ * TODO : answer to above from C-API:
+ *  Note: This function can only produce objects, not values. Use the
+ *       `realm_results_t` returned by `realm_query_find_all()` to retrieve
+ *        values from a list of primitive values.
+ */
+interface RealmSingleQuery<E> : Flowable<E> {
+
+    /**
+     * Finds the first object that fulfills the query conditions and returns it in a blocking
+     * fashion.
+     *
+     * Launching heavy queries from the UI thread may result in a drop of frames or even ANRs. **We
+     * do not recommend doing so.** If you want to prevent these behaviors you can obtain the
+     * object asynchronously using [asFlow] instead.
+     *
+     * @return a [RealmObject] instance or `null` if no object matches the condition.
```

Comment:
There is some corner cases around this that might turn out slightly problematic, especially in the type system.

1. We are going to allow queries of e.g. `RealmList<Integer>` so other things than RealmObjects can be returned.
2. When we add fine-grained notifications, we will have a problem as e.g. `ObjectChange<E>` requires `E` to be a a subclass of RealmObject... I'm still not 100% sure how we can resolve this.